pkg/ecdsa/dkg/states.go

```diff
@@ -107,35 +134,290 @@ func (skgs *symmetricKeyGenerationState) Receive(msg net.Message) error {
+	switch phaseMessage := msg.Payload().(type) {
+	case *tssRoundOneMessage:
+		if tros.member.shouldAcceptMessage(
+			phaseMessage.SenderID(),
+			msg.SenderPublicKey(),
+		) && tros.member.sessionID == phaseMessage.sessionID {
+			tros.phaseMessages = append(tros.phaseMessages, phaseMessage)
+		}
+	}
+
+	return nil
+}
+
+func (tros *tssRoundOneState) Next() state.State {
+	return &tssRoundTwoState{
+		channel:               tros.channel,
+		member:                tros.member.initializeTssRoundTwo(),
+		previousPhaseMessages: tros.phaseMessages,
+	}
+}
+
+func (tros *tssRoundOneState) MemberIndex() group.MemberIndex {
+	return tros.member.id
+}
+
+// tssRoundOneState is the state during which members broadcast TSS
+// shares and de-commitments.
+// `tssRoundTwoMessage`s are valid in this state.
+type tssRoundTwoState struct {
+	channel net.BroadcastChannel
+	member  *tssRoundTwoMember
+
+	previousPhaseMessages []*tssRoundOneMessage
+
+	phaseMessages []*tssRoundTwoMessage
+}
+
+func (trts *tssRoundTwoState) DelayBlocks() uint64 {
+	return tssRoundTwoStateDelayBlocks
+}
+
+func (trts *tssRoundTwoState) ActiveBlocks() uint64 {
+	return tssRoundTwoStateActiveBlocks
+}
+
+func (trts *tssRoundTwoState) Initiate(ctx context.Context) error {
+	trts.member.MarkInactiveMembers(trts.previousPhaseMessages)
+
+	if len(trts.member.group.OperatingMemberIDs()) != trts.member.group.GroupSize() {
+		return fmt.Errorf("inactive members detected")
+	}
+
+	// The ctx instance passed as Initiate argument is scoped to the lifetime
+	// of the current state. However, the Initiate method is blocking and the
+	// ctx instance is cancelled properly only after Initiate returns. Because
+	// of that, we cannot use ctx as round timeout signal as Initiate would
+	// hang forever if something goes wrong. To avoid such a resource leak,
+	// we set a round timeout based on state block duration and an average
+	// block time. The exact duration doesn't need to be super-accurate because
+	// if the timeout is hit, the execution will fail anyway. We just want
+	// to give enough time for round computation and make sure the round
+	// terminates regardless of the result.
+	stateBlocks := trts.DelayBlocks() + trts.ActiveBlocks()
+	stateDuration := 15 * time.Duration(stateBlocks)
+	roundCtx, roundCtxCancel := context.WithTimeout(ctx, stateDuration)
+	defer roundCtxCancel()
+
+	message, err := trts.member.tssRoundTwo(roundCtx, trts.previousPhaseMessages)
+	if err != nil {
+		return err
+	}
+
+	if err := trts.channel.Send(ctx, message); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (trts *tssRoundTwoState) Receive(msg net.Message) error {
+	switch phaseMessage := msg.Payload().(type) {
+	case *tssRoundTwoMessage:
+		if trts.member.shouldAcceptMessage(
+			phaseMessage.SenderID(),
+			msg.SenderPublicKey(),
+		) && trts.member.sessionID == phaseMessage.sessionID {
+			trts.phaseMessages = append(trts.phaseMessages, phaseMessage)
+		}
+	}
+
+	return nil
+}
```

Comment:
is this function capable of producing an error?

Reply:
ah, we probably need it to return an `error` anyway to make it fit in the `Receive` api

Reply:
Yep, we don't produce an error but we must implement the `state.State` interface.